### ISSUE TITLE
Automatically generate Virtual Network entry

### ISSUE THREAD

**Issue by nikosT** [EIDA/routing#43]
Since the `routing.xml` file contains explicitly all of each stations and due to the fact that Virtual Networks can consist of networks that belong to different nodes (e.g. _NFOCRL), there should be a way to solve add/removals of stations that occurred daily (e.g. new stations have just been added and vice versa).

I would suggest that all involved nodes should implement the corresponding entries to the `routing.xml`, in a similar way, as they were the ones that containing the virtual network. However, there would be a tag that indicates the "master" node and the "slaves" ones, for the specific virtual network. Daily the "master" node, will collect the changes to its `routing.xml` by requesting all other nodes' routing service [or only the "slaves" ones, but this means that master node needs to know a list of the other involved nodes].

For example, let's say we have:

Node no1:
Network: A
Stations: A1,A2

Node no2:
Network: B
Stations: B1,B2

and assuming that we have virtual network _X on node no2 which consist of A2, B1, B2.
So the master node is no2.

Thus,
node no1 will have:
``` xml
<ns0:vnetwork networkCode="_X" state="slave">
<ns0:stream networkCode="A" stationCode="A2" locationCode="*" streamCode="*" start="2018-01-01T00:00:00" end=""/>
```

node no2 will have:
``` xml
<ns0:vnetwork networkCode="_X" state="master">
<ns0:stream networkCode="B" stationCode="B1" locationCode="*" streamCode="*" start="2018-01-01T00:00:00" end=""/>
<ns0:stream networkCode="B" stationCode="B2" locationCode="*" streamCode="*" start="2018-01-01T00:00:00" end=""/>
```
Now,
Daily node no1 will do nothing, since there is no "master" tag. On the other hand node no2 will request by using the routing service of each node to collect the stations that involded to specific virtual network, and will implement a new virtual network entry but without the state (actually exactly as it is right now), that contains all stations (i.e.):
``` xml
<ns0:vnetwork networkCode="_X">
<ns0:stream networkCode="A" stationCode="A2" locationCode="*" streamCode="*" start="2018-01-01T00:00:00" end=""/>
<ns0:stream networkCode="B" stationCode="B1" locationCode="*" streamCode="*" start="2018-01-01T00:00:00" end=""/>
<ns0:stream networkCode="B" stationCode="B2" locationCode="*" streamCode="*" start="2018-01-01T00:00:00" end=""/>
```
The above tag `<ns0:vnetwork networkCode="_X">` (and what it contains) would be re-created daily -only to the master node ofcourse-.

Since it would be daily it wouldn't harm the performance. But this can achieve a better one, if its virtual network maintains a list of the involved nodes (for the particular virtual network). This however means, that the node operators need to update the list of nodes for the specific network, if there is any change to this -probably not that often-.

In that case would have:
node no1:
``` xml
<ns0:vnetwork networkCode="_X">
<ns0:stream networkCode="A" stationCode="A2" locationCode="*" streamCode="*" start="2018-01-01T00:00:00" end=""/>
```
node no2:
``` xml
<ns0:vnetwork networkCode="_X" list="http://nodeno1"> #comma separated
<ns0:stream networkCode="A" stationCode="A2" locationCode="*" streamCode="*" start="2018-01-01T00:00:00" end=""/>
```
and the routing service to the master node would still generate daily the full entry:
``` xml
<ns0:vnetwork networkCode="_X">
<ns0:stream networkCode="A" stationCode="A2" locationCode="*" streamCode="*" start="2018-01-01T00:00:00" end=""/>
<ns0:stream networkCode="B" stationCode="B1" locationCode="*" streamCode="*" start="2018-01-01T00:00:00" end=""/>
<ns0:stream networkCode="B" stationCode="B2" locationCode="*" streamCode="*" start="2018-01-01T00:00:00" end=""/>
```
To sum up:
1. All involved nodes have a virtual network entry in state "slave" and including ONLY their stations to it (concerning this particular virtual network).
2. The primary node will do exact the step 1, apart from setting the state to "master".

(Also the primary node will observe that daily there would be an automatically generation of `<ns0:vnetwork networkCode="_X">` tag entry that includes all involved stations in its `routing.xml`)

**Comment by Jollyfant** [comment]
Maybe a stupid question but how is it different from a non-virtual network such as Z3 that is distributed across nodes?

**Comment by nikosT** [comment]
Actually it could be completely distributed, without even master/slave. For instance, every node adds its virtual network entry with each stations. Routing service loops over all nodes to merge stations from same virtual network or redirect to real stations anyway.

**Comment by javiquinte** [comment]
I think it could be interesting, but we need more participation *and* consensus from the other members.

**Comment by kaestli** [comment]
Note that the current routing specification is not explicit on this point - neither for the physical network nor for the virtual network. As we have a network 4C at one node with some stations and a network 4C at another node with some other stations (and we mean that together they form the virtual network 4C), we could do so on virtual networks.
Both cases (having the definition of one virtual network with physical stations of different endpoints defined at one endpoint only, or having two partial definitions at two endpoints) may make sense:
- if it is a joint Vnetwork of two institutions, distributed definitions may ease maintenance (track only local changes in the underlying physical networks)
- if it is a Vnetwork for a purpose defined by one institution only, it should be defined by this institution only, even if it includes some foreign stations.

A further fuzzy point in the specification is whether wildcards are allowed - the example given does not contain them, but this does not rule it out that they may be allowed.

There are use cases where this would make sense.
You could , e.g., easily define the "European Very Broad Band Virtual Network" as follows:
```
<ns0:vnetwork networkCode="_EUROVBBNET">
<ns0:stream networkCode="*" stationCode="*" locationCode="*" streamCode="HH?" start="1980-01-01T00:00:00" end=""/>
</ns0:vnetwork>
```
... and the definition would even keep up with new networks coming into existence and new stations being deployed.

The federator can handle all of these cases: distributed vnetwork definitions as well as non-distributed ones including stations at other end points, and explicit includes as well as wild cards on any SNCL components. 
Still, it would be welcome to have explicit statements on the available options in the specs document, and an xsd or relaxng schema for the localconfig xml.
